### ISSUE TITLE
core::slice::Iter::new is unsound for non-ZSTs

### ISSUE THREAD

**Issue by CoffeeBlend** [rust-lang/rust#83996]
The current implementation uses `pointer::add` to compute the end pointer for the bounds check:
https://github.com/rust-lang/rust/blob/69e1d22ddbc67b25141a735a22a8895a678b32ca/library/core/src/slice/iter.rs#L88-L102
The method requires that the calculation will not overflow a `usize`, however that is not always the case. For instance, an allocator might return the last available page (`0xfffff000` on x86) and correctly return a slice of `u8` (with size 4096 on x86). If a program now iterates over the slice, the end pointer will overflow, wrapping around the address space and thus creating UB. 

This behaviour is extremely unlikely and only occurs with `no_std` as most kernels reserve the higher half of the address space anyway. 

# Solutions
* Use wrapping_add instead, which avoids the UB, but might disable some optimizations
* Update the requirements for allocators so they aren't allowed to return the last bit of memory
* (Ignore the problem, it is extremely unlikely to cause problems after all)

I might be wrong about the nature of this issue, I know little about LLVM and whether this is actually UB or just a flaw in the documentation.

**Comment by the8472** [comment]
`vec::IntoIter` is constructed in a similar fashion

https://github.com/rust-lang/rust/blob/010c2368fa33ef08df3080228498db8d2ba5a93b/library/alloc/src/vec/mod.rs#L2429-L2439

`array::IntoIter` on the other hand uses a different idiom

https://github.com/rust-lang/rust/blob/010c2368fa33ef08df3080228498db8d2ba5a93b/library/core/src/array/iter.rs#L26-L33

**Comment by RalfJung** [comment]
> The method requires that the calculation will not overflow a usize, however that is not always the case. For instance, an allocator might return the last available page (0xfffff000 on x86) and correctly return a slice of u8 (with size 4096 on x86). If a program now iterates over the slice, the end pointer will overflow, wrapping around the address space and thus creating UB.

An allocator must ensure that the one-past-the-end address of an allocation does *not* overflow. In other words, `0xffffffff` can never be inside an allocation, it can only be one-past-the-end of an allocation. I think this rules out your example, doesn't it?

**Comment by CoffeeBlend** [comment]
Yes, that would solve the issue! I can't seem to find it in the documentation, however. Should I change the labeling of this issue to make it a doc-bug?

**Comment by RalfJung** [comment]
Yeah this could certainly be documented better.

**Comment by scottmcm** [comment]
I think this has always been part of the implicit invariant for slices, but it would definitely be good to mention it explicitly in things like the documentation for https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety  (The last bullet there is coming from the expectation that `data.offset(len)` needs to be valid, it doesn't seem to be directly mentioned.)

**Comment by RalfJung** [comment]
This is not really specific to slices though, it is a property of all "allocated objects" in Rust and the allocator must comply by it.

**Comment by m-ou-se** [comment]
We assume in lots of places that `slice.as_ptr().add(slice.len())` is safe. E.g. https://github.com/rust-lang/rust/blob/d408fdd4a82bc3e7ea61dd81bc9a8781b2bf939d/library/core/src/slice/mod.rs#L496-L517

The documentation for `pointer::add` does mention that `vec.as_ptr().add(vec.len())` is always safe, but doesn't state that clearly for slices in general: https://github.com/rust-lang/rust/blob/d408fdd4a82bc3e7ea61dd81bc9a8781b2bf939d/library/core/src/ptr/const_ptr.rs#L195-L196

The implementation of `Index` for slices also assumes this is always safe: https://github.com/rust-lang/rust/blob/d408fdd4a82bc3e7ea61dd81bc9a8781b2bf939d/library/core/src/slice/index.rs#L166-L170

**Comment by the8472** [comment]
Should that part of the allocator API contracts?

**Comment by apiraino** [comment]
Removing the `I-prioritize` as it looks like more relevant to documentation
(but in case apply a priority if it helps a follow-up)

@rustbot label -I-prioritize

**Comment by RalfJung** [comment]
> Should that part of the allocator API contracts?

Yes, this effectively already is part of the allocator API contract.
